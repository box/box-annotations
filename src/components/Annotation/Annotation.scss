@import '../../common.scss';

.ba .ba-annotation {
    @import '~box-react-ui/lib/styles/common/inline-notifications';

    position: relative;
    z-index: 1;

    &.ba-is-pending {
        opacity: 50%;
    }

    .ba-annotation-content {
        display: flex;
        min-width: 0;
        width: 100%;
    }

    .ba-annotation-avatar {
        flex-shrink: 0;
        margin-right: 17px; // Spaces avatar circle from text to the right
        margin-top: 3px; // Required to top align avatar circle with text to the right
    }

    .bcs-comment-text {
        white-space: pre-wrap;
        word-wrap: break-word;
    }

    .bcs-comment-error {
        display: block;
        padding: 15px;
    }

    .bcs-comment-error-action {
        color: $king-crimson;
        display: block;
        margin-top: 5px; // Required to distinguish between action and error text
        text-decoration: underline;
    }
}

.ba-mobile-annotation-dialog .ba-annotation {
    .bcs-comment-text {
        font-size: 15px;
        width: 100%;
    }
}
