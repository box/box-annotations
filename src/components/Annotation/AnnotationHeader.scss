@import '../../variables.scss';

.ba .ba-annotation-headline {
    display: flex;
    flex-flow: row wrap;

    &:hover,
    &.ba-is-focused {
        .bcs-comment-delete-container {
            opacity: 1;
        }
    }

    .ba-annotation-avatar {
        flex-shrink: 0;
        margin-right: 17px; // Spaces avatar circle from text to the right
        margin-top: 3px; // Required to top align avatar circle with text to the right
    }

    .ba-annotation-user-name {
        color: $twos;
        font-weight: bold;
        margin-right: 10px; // Spaces user name from timestamp
    }

    .ba-annotation-created-at {
        color: $nines;
        cursor: default;
        font-size: smaller;
    }

    .bcs-comment-delete-container {
        margin-top: 3px; // Required to top align trash icon with text to the left
        opacity: 0;
        position: absolute;
        right: 0;
        top: 0;
        z-index: 1;

        &.ba-is-confirming {
            opacity: 1;
        }
    }

    .bcs-comment-delete {
        vertical-align: top;
    }
}

.flyout-overlay {
    .bcs-comment-delete-yes,
    .bcs-comment-delete-no {
        font-size: inherit;
    }
}

.ba-mobile-annotation-dialog .ba-annotation-headline {
    .ba-annotation-avatar {
        width: 43px;

        img {
            height: 40px;
            width: 40px;
        }
    }

    .ba-annotation-user-name {
        font-size: 16px;
    }

    .bcs-comment-delete-container {
        font-size: 15px;
        opacity: 1;
    }

    .bcs-comment-delete svg {
        height: 18px;
        width: 18px;
    }
}
