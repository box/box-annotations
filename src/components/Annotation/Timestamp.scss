@import '../../variables.scss';

.ba .ba-annotation-created-at {
    color: $nines;
    cursor: default;
    font-size: smaller;
}

.ba-mobile-annotation-dialog .ba-annotation-created-at {
    font-size: 14px;
}
