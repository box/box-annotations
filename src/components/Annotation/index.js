// @flow
export { default } from './Annotation';
