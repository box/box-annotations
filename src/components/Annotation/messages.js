/* eslint-disable max-len */
/**
 * @flow
 * @file i18n messages
 * @author Box
 */

import { defineMessages } from 'react-intl';
import type { MessageDescriptor } from 'react-intl';

const messages: { [string]: MessageDescriptor } = defineMessages({
    loadErrorMessage: {
        id: 'ba.loadErrorMessage',
        description: 'Error when annotations fail to load',
        defaultMessage: 'We\'re sorry, annotations failed to load for this file.'
    },
    authErrorMessage: {
        id: 'ba.authErrorMessage',
        description: 'Error when the user\'s token has expired',
        defaultMessage: 'Your session has expired. Please refresh the page.'
    },
    posting: {
        id: 'ba.posting',
        description: 'Placeholder when the annotation is pending',
        defaultMessage: 'Posting...'
    },
    anonymousUserName: {
        id: 'ba.anonymousUserName',
        description: 'Placeholder when the current annotation\'s user information is unknown',
        defaultMessage: 'Some User'
    },
    pointToggle: {
        id: 'ba.pointToggle',
        description: 'Point annotation mode toggle',
        defaultMessage: 'Point annotation mode'
    },
    highlightToggle: {
        id: 'ba.highlightToggle',
        description: 'Plain highlight annotation toggle',
        defaultMessage: 'Highlight tex'
    },
    highlightCommentToggle: {
        id: 'ba.highlightComment',
        description: 'Highlight comment annotation toggle',
        defaultMessage: 'Add comment to highlighted text'
    },
    whoHighlighted: {
        id: 'ba.whoHighlighted',
        description: 'Label for who highlighted the annotated text',
        defaultMessage: '{1} highlighted'
    },
    drawToggle: {
        id: 'ba.drawToggle',
        description: 'Drawing annotation mode toggle',
        defaultMessage: 'Drawing annotation mode'
    },
    whoDrew: {
        id: 'ba.whoDrew',
        description: 'Label for who drew the drawing annotation',
        defaultMessage: '{1} drew'
    }
});

export default messages;
