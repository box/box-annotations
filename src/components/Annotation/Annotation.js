// @flow
import * as React from 'react';
import classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import getProp from 'lodash/get';
import noop from 'lodash/noop';

import Avatar from 'box-react-ui/lib/components/avatar';
import { ReadableTime } from 'box-react-ui/lib/components/time';
import Tooltip from 'box-react-ui/lib/components/tooltip';

import CommentText from '../../../third-party/components/CommentText';
import InlineDelete from '../../../third-party/components/InlineDelete';
import CommentInlineError from '../../../third-party/components/CommentInlineError';
import UserLink from '../../../third-party/components/UserLink';
import Internationalize from '../Internationalize';
import withFocus from '../withFocus';
import messages from './messages';

import './Annotation.scss';

const ONE_HOUR_MS = 3600000; // 60 * 60 * 1000

type Props = {
    id: string,
    message: string,
    permissions: AnnotationPermissions,
    createdAt?: string,
    createdBy?: User,
    modifiedAt?: string,
    onDelete?: Function,
    isPending?: boolean,
    error?: ActionItemError,
    language?: string,
    messages?: StringMap,
    className: string,
    onBlur: Function,
    onFocus: Function,
    locale: string
};

class Annotation extends React.PureComponent<Props> {
    static defaultProps = {
        isPending: false,
        onDelete: noop
    };

    render() {
        const {
            id,
            isPending,
            error,
            createdAt,
            createdBy,
            permissions,
            message,
            onDelete,
            className,
            onBlur,
            onFocus,
            language,
            messages: intlMessages
        } = this.props;

        const canDelete = getProp(permissions, 'can_delete', false);
        const hasDeletePermission = onDelete && canDelete && !isPending;

        const createdAtTimestamp = createdAt ? new Date(createdAt).getTime() : null;

        const annotator = {
            id: createdBy && createdBy.name ? createdBy.id : '0',
            name:
                createdBy && createdBy.name ? (
                    createdBy.name
                ) : (
                    <FormattedMessage className='ba-annotation-user-name' {...messages.anonymousUserName} />
                )
        };

        return (
            <Internationalize language={language} messages={intlMessages}>
                <div
                    className={classNames(`ba-annotation ${className}`, {
                        'ba-is-pending': isPending || error
                    })}
                    onBlur={onBlur}
                    onFocus={onFocus}
                >
                    <div className='ba-annotation-content'>
                        <Avatar className='ba-annotation-avatar' {...createdBy} />
                        <div className='ba-annotation-text'>
                            <div className='ba-annotation-headline'>
                                <UserLink className='ba-annotation-user-name' {...annotator} />
                                {createdAtTimestamp && (
                                    <Tooltip
                                        text={
                                            <FormattedMessage
                                                {...messages.annotationPostedFullDateTime}
                                                values={{ time: createdAtTimestamp }}
                                            />
                                        }
                                    >
                                        <time className='ba-annotation-created-at'>
                                            <ReadableTime
                                                timestamp={createdAtTimestamp}
                                                relativeThreshold={ONE_HOUR_MS}
                                            />
                                        </time>
                                    </Tooltip>
                                )}
                                {hasDeletePermission ? (
                                    <InlineDelete
                                        id={id}
                                        permissions={permissions}
                                        message={<FormattedMessage {...messages.annotationDeletePrompt} />}
                                        onDelete={onDelete}
                                    />
                                ) : null}
                            </div>
                            <CommentText id={id} tagged_message={message} translationEnabled={false} />
                        </div>
                    </div>
                    {error ? <CommentInlineError {...error} /> : null}
                </div>
            </Internationalize>
        );
    }
}

export { Annotation as AnnotationComponent };
export default withFocus(Annotation);
