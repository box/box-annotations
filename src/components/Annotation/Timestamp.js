// @flow
import * as React from 'react';
import { FormattedMessage } from 'react-intl';

import { ReadableTime } from 'box-react-ui/lib/components/time';
import Tooltip from 'box-react-ui/lib/components/tooltip';

import messages from './messages';
import './Timestamp.scss';

const ONE_HOUR_MS = 3600000; // 60 * 60 * 1000

type Props = {
    time: string
};

const Timestamp = ({ time }: Props) => {
    const createdAtTimestamp = new Date(time).getTime();

    return (
        <Tooltip
            text={<FormattedMessage {...messages.annotationPostedFullDateTime} values={{ time: createdAtTimestamp }} />}
        >
            <time className='ba-annotation-created-at'>
                <ReadableTime timestamp={createdAtTimestamp} relativeThreshold={ONE_HOUR_MS} />
            </time>
        </Tooltip>
    );
};

export default Timestamp;
