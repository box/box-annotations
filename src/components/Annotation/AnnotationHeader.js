// @flow
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import getProp from 'lodash/get';

import InlineDelete from '../../../third-party/components/Comment/InlineDelete';
import UserLink from '../../../third-party/components/Comment/UserLink';
import Timestamp from './Timestamp';
import messages from './messages';

import './AnnotationHeader.scss';

const ANONYMOUS_USER = {
    id: '0',
    name: <FormattedMessage className='ba-annotation-user-name' {...messages.anonymousUserName} />
};

type Props = {
    id: string,
    permissions: AnnotationPermissions,
    createdAt?: string,
    createdBy?: User,
    onDelete?: Function,
    isPending: boolean
};

const AnnotationHeader = ({ id, permissions, onDelete, createdAt, createdBy, isPending }: Props) => {
    const canDelete = getProp(permissions, 'can_delete', false);
    const hasDeletePermission = onDelete && canDelete && !isPending;

    const annotator = !!createdBy && !!createdBy.name ? createdBy : ANONYMOUS_USER;

    return (
        <div className='ba-annotation-headline'>
            <UserLink className='ba-annotation-user-name' id={annotator.id} name={annotator.name} />
            {createdAt && <Timestamp time={createdAt} />}
            {hasDeletePermission && (
                <InlineDelete
                    id={id}
                    permissions={permissions}
                    message={<FormattedMessage {...messages.annotationDeletePrompt} />}
                    onDelete={onDelete}
                />
            )}
        </div>
    );
};

export default AnnotationHeader;
