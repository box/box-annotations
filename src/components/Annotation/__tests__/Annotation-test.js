import * as React from 'react';
import { mount, shallow } from 'enzyme';

import { FormattedMessage } from 'react-intl';
import Annotation from '../Annotation';
import messages from '../messages';

const TIME_STRING_SEPT_27_2017 = '2017-09-27T10:40:41-07:00';

const USER = {
    type: 'user',
    id: '789ghi',
    name: 'Mia Thermopolis',
    email: 'princess@genovia.gov'
};

const annotation = {
    id: '123',
    createdAt: TIME_STRING_SEPT_27_2017,
    createdBy: USER,
    message: 'test',
    permissions: {}
};

describe('components/Annotation', () => {
    beforeEach(() => {});

    const render = (props = {}) => shallow(<Annotation {...annotation} {...props} />).dive();
    const deepRender = (props = {}) => mount(<Annotation {...annotation} {...props} />);

    test('should correctly render annotation', () => {
        const unixTime = new Date(TIME_STRING_SEPT_27_2017).getTime();

        const wrapper = render();

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('ReadableTime').prop('timestamp')).toEqual(unixTime);
        expect(wrapper.find('UserLink').prop('name')).toEqual(USER.name);
    });

    test('should correctly display anonymous username if no name was provided in createdBy', () => {
        const unknownUser = {
            type: 'user',
            id: '0'
        };
        const anonymousUserName = (
            <FormattedMessage className='ba-annotation-user-name' {...messages.anonymousUserName} />
        );
        const wrapper = render({ createdBy: unknownUser });
        expect(wrapper.find('UserLink').prop('name')).toEqual(anonymousUserName);
    });

    test('should not allow actions when annotation is pending', () => {
        const wrapper = render({
            onDelete: jest.fn(),
            permissions: { can_delete: true },
            isPending: true
        });
        expect(wrapper.find('InlineDelete').length).toEqual(0);
    });

    // eslint-disable-next-line
    test('should allow user to delete if they have delete permissions on the annotation and delete handler is defined', () => {
        const wrapper = render({ onDelete: jest.fn(), permissions: { can_delete: true } });
        expect(wrapper.find('InlineDelete').length).toEqual(1);
    });

    test('should not allow user to delete if they lack delete permissions on the annotation', () => {
        const wrapper = render({ onDelete: jest.fn() });
        expect(wrapper.find('InlineDelete').length).toEqual(0);
    });

    test('should not allow user to delete if onDelete handler is undefined', () => {
        const wrapper = render();
        expect(wrapper.find('InlineDelete').length).toEqual(0);
    });

    test('should render an error when one is defined', () => {
        const wrapper = render({
            onDelete: jest.fn(),
            error: {
                title: 'error',
                message: 'errorrrrr'
            }
        });

        expect(wrapper).toMatchSnapshot();
    });

    test('should render an error cta when an action is defined', () => {
        const onActionSpy = jest.fn();

        const wrapper = deepRender({
            onDelete: jest.fn(),
            error: {
                title: 'error',
                message: 'errorrrrr',
                action: {
                    text: 'click',
                    onAction: onActionSpy
                }
            }
        });

        const inlineErrorActionLink = wrapper.find('InlineError').find('PlainButton.lnk');
        expect(inlineErrorActionLink.length).toEqual(1);

        inlineErrorActionLink.simulate('click');

        expect(onActionSpy).toHaveBeenCalledTimes(1);
    });
});
