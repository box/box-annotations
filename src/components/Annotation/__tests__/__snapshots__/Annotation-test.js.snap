// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Annotation should correctly render annotation 1`] = `
<Internationalize>
  <div
    className="ba-annotation "
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      className="ba-annotation-content"
    >
      <Avatar
        className="ba-annotation-avatar"
        email="princess@genovia.gov"
        id="789ghi"
        name="Mia Thermopolis"
        type="user"
      />
      <div
        className="ba-annotation-text"
      >
        <div
          className="ba-annotation-headline"
        >
          <UserLink
            className="ba-annotation-user-name"
            id="789ghi"
            name="Mia Thermopolis"
          />
          <Tooltip
            constrainToScrollParent={false}
            constrainToWindow={true}
            position="top-center"
            text={
              <FormattedMessage
                defaultMessage="{time, date, full} at {time, time, short}"
                id="ba.annotationPostedFullDateTime"
                values={
                  Object {
                    "time": 1506534041000,
                  }
                }
              />
            }
            theme="default"
          >
            <time
              className="ba-annotation-created-at"
            >
              <ReadableTime
                relativeThreshold={3600000}
                timestamp={1506534041000}
              />
            </time>
          </Tooltip>
        </div>
        <CommentText
          id="123"
          tagged_message="test"
          translationEnabled={false}
        />
      </div>
    </div>
  </div>
</Internationalize>
`;

exports[`components/Annotation should render an error when one is defined 1`] = `
<Internationalize>
  <div
    className="ba-annotation  ba-is-pending"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      className="ba-annotation-content"
    >
      <Avatar
        className="ba-annotation-avatar"
        email="princess@genovia.gov"
        id="789ghi"
        name="Mia Thermopolis"
        type="user"
      />
      <div
        className="ba-annotation-text"
      >
        <div
          className="ba-annotation-headline"
        >
          <UserLink
            className="ba-annotation-user-name"
            id="789ghi"
            name="Mia Thermopolis"
          />
          <Tooltip
            constrainToScrollParent={false}
            constrainToWindow={true}
            position="top-center"
            text={
              <FormattedMessage
                defaultMessage="{time, date, full} at {time, time, short}"
                id="ba.annotationPostedFullDateTime"
                values={
                  Object {
                    "time": 1506534041000,
                  }
                }
              />
            }
            theme="default"
          >
            <time
              className="ba-annotation-created-at"
            >
              <ReadableTime
                relativeThreshold={3600000}
                timestamp={1506534041000}
              />
            </time>
          </Tooltip>
        </div>
        <CommentText
          id="123"
          tagged_message="test"
          translationEnabled={false}
        />
      </div>
    </div>
    <CommentInlineError
      message="errorrrrr"
      title="error"
    />
  </div>
</Internationalize>
`;
